### PR TITLE
Variadic executor parameters

### PR DESCRIPTION
**Pull request by mcopik**
This is my first attempt to provide a syntax for multiple executor parameters passed through a variadic template. I'm not completely happy with this solution, maybe there is a better way to implement unwrapping, I tried to find a way with the least amount of code duplication (which is already an unsolvable problem in execution policies).

The implementation creates a single parameter through variadic inheritance, which obviously prohibits duplication of parameter types; an additional static_assert in constructor provides a better error information than a compiler complaining about ambiguous choice between functions. Another possible option could be to use boost::hana (since 1.61) and implement it using compile-time algorithms on a tuple. I don't see any other way.

The biggest trick was to implement a support for parameters passed as a reference wrapper (Boost or C++11). For that purpose we need an unwrapper which API exposes only methods defined in the wrapped parameter; solution similar to executor_parameter_traits would cause another name clash between different parameters. Unfortunately, the problem is defining functions at type construction, not function overload, hence we can't use SFINAE directly. Just using a auto and decltype() with a fake template parameter is not sufficient. 

My solution uses std::enable_if and creates different parameter types, where a tag is bound to specific methods; if executor parameters inherits from chunk tag, chunk size methods will be exposed etc. I couldn't use parameter traits directly, because it will always provide a default behaviour if a specific function doesn't exist -> duplication. The small benefit of this solution is clear ordering and documenting different types of parameters. 

I have found one solution which doesn't use separate tags: for every method a checker similar to traits is implemented - design is the same, but instead of running either the parameter function or default case, a constexpr boolean is exposed (true iff a method is present in type). But then we would have two huge code bases doing almost the same thing. Merging it requires rewriting whole traits, removing helpers and dispatching functions with an enable_if. I could implement this solution, but it seems that it would be a huge change in existing code base.

### CODE REVIEW

File: hpx/parallel/execution_policy.hpp [hunk 1]
```diff
@@ -162,24 +163,23 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
...
         ///
-        /// \note Requires: is_executor_parameters<Parameters>::value is true
+        /// \note Requires: all parameters are executor_parameters,
+        ///                 different parameter types can't be duplicated
         ///
         /// \returns The new sequential_task_execution_policy
         ///
-        template <typename Parameters>
+        template <typename... Parameters, typename ParametersType =
+            typename hpx::parallel::executor_parameters_join<Parameters...>::type>
```
Comment (hkaiser) [inline hpx/parallel/execution_policy.hpp]: What's the rationale for making the `ParametersType` a template argument? Would it be sufficient to directly use the `hpx::parallel::executor_parameters_join<Parameters...>` in the body of this function?

File: hpx/parallel/execution_policy.hpp [hunk 2]
```diff
@@ -563,24 +561,23 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
...
         ///
-        template <typename Parameters_>
+        template <typename... Parameters_, typename ParametersType =
+            typename hpx::parallel::executor_parameters_join<Parameters_...>::type>
         typename rebind_executor<
-            sequential_execution_policy_shim, executor_type, Parameters_
+            sequential_execution_policy_shim, executor_type, ParametersType
         >::type
-        with(Parameters_& params) const
+        with(Parameters_ &&... params) const
```
Comment (hkaiser) [inline hpx/parallel/execution_policy.hpp]: Good catch, thanks!

File: hpx/parallel/execution_policy.hpp [hunk 3]
```diff
@@ -162,24 +163,23 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
...
         typename rebind_executor<
-            sequential_task_execution_policy, executor_type, Parameters
+            sequential_task_execution_policy, executor_type, ParametersType
         >::type
-        with(Parameters && params) const
+        with(Parameters &&... params) const
         {
-            static_assert(
-                is_executor_parameters<Parameters>::value,
-                "is_executor_parameters<Parameters>::value");
```
Comment (hkaiser) [inline hpx/parallel/execution_policy.hpp]: Shouldn't we keep the verification in place that all `Parameters...` types are in fact executor parameters?

File: hpx/parallel/executors/auto_chunk_size.hpp [hunk 4]
```diff
@@ -55,6 +55,12 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v3)
           : min_time_(rel_time.value().count())
         {}
 
+        template <typename Executor>
+        static bool variable_chunk_size(Executor&)
+        {
+            return false;
+        }
```
Comment (hkaiser) [inline hpx/parallel/executors/auto_chunk_size.hpp]: Why is this necessary? The `executor_parameters_traits` should supply a default implementation of this function, if needed.

File: hpx/parallel/executors/executor_parameters.hpp [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+#include <boost/ref.hpp>
+namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v3)
+{
+    namespace detail {
+
+        template<template<typename> class Condition, typename Arg,
+            typename Enable = void>
+        struct counter_increment
+        {
+            constexpr static int value = 0;
```
Comment (hkaiser) [inline hpx/parallel/executors/executor_parameters.hpp]: Please use `HPX_CONSTEXPR` instead of `constexpr`. SOme compilers we support don't handle `constexpr` correctly yet.

File: hpx/parallel/executors/executor_parameters.hpp [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+            template<typename U>
+            unwrapper(U && u) : T(u) {}
+        };
+
+        template<typename T>
+        struct unwrapper< boost::reference_wrapper<T> >
+        {
+            template<typename WrapperType>
+            unwrapper(WrapperType && wrap) : wrap(wrap) {}
```
Comment (hkaiser) [inline hpx/parallel/executors/executor_parameters.hpp]: Shouldn't this be: `... : wrap(std::forward<WrapperType>(wrap))`? Also, please use distinct names for members and arguments.

File: hpx/parallel/executors/executor_parameters.hpp [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+            template<typename Executor, typename U = T, typename std::enable_if<
+                is_executor_parameters_chunk_size<U>::value >::type* = nullptr
+                >
```
Comment (hkaiser) [inline hpx/parallel/executors/executor_parameters.hpp]: We have not agreed on using `nullptr` yet, sorry. 

@K-ballo would it be safe to assume that it's supported on all of our platforms? 

ALternatively, and if you want to use it, please add a feature test verifying whether it is supported on a particular platform. Also, please introduce a macro `HPX_NULLPTR` which falls back to `0` if it's not supported.

File: hpx/traits/is_executor_parameters.hpp [hunk 8]
```diff
@@ -23,37 +23,83 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v3)
...
+    struct executor_parameters_processing_units_count_tag : executor_parameters_tag {};
+
+    /// Parameters implement function: void reset_thread_distribution(Executor &)
+    struct executor_parameters_reset_thread_distr_tag : executor_parameters_tag {};
 
     namespace detail
     {
         /// \cond NOINTERNAL
-        template <typename T>
+        template <typename T, typename Tag>
```
Comment (hkaiser) [inline hpx/traits/is_executor_parameters.hpp]: Should this be changed to: `template <typename T, typename Tag = executor_parameters_tag>` ?

File: hpx/traits/is_executor_parameters.hpp [hunk 9]
```diff
@@ -23,37 +23,83 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v3)
...
+        >
+    {};
+
+    template <typename T>
+    struct is_executor_parameters_processing_units_count
+      : detail::is_executor_parameters<
+            typename hpx::util::decay<T>::type,
+            executor_parameters_processing_units_count_tag
+        >
     {};
```
Comment (hkaiser) [inline hpx/traits/is_executor_parameters.hpp]: This multitude of tags requires for a parameters object to be derived from more than one of those, eventually... Is that really a proper solution?

File: hpx/traits/is_executor_parameters.hpp [hunk 10]
```diff
@@ -67,6 +113,26 @@ namespace hpx { namespace traits
     struct is_executor_parameters
       : parallel::v3::is_executor_parameters<Parameters>
     {};
+
+    template <typename Parameters, typename Enable>
```
Comment (hkaiser) [inline hpx/traits/is_executor_parameters.hpp]: Should the Enable parameter be defaulted to `= void`?

File: hpx/parallel/execution_policy.hpp (hunk 1 above)
Comment (mcopik) [inline hpx/parallel/execution_policy.hpp]: Of course we can, it's just how I've been using it all the time. I got used to this solution, because this allows me to use it easily in specifying type before function body or in inheritance. Hence I do not need to declare this type twice, one for return type of function and then again in the function body. 

Is it considered a bad practice, bad style? Should I change it? I thought that it's better to have it as a default parameter rather than duplicate code, just like it happens in rebound_type.

File: hpx/parallel/execution_policy.hpp (hunk 3 above)
Comment (mcopik) [inline hpx/parallel/execution_policy.hpp]: I moved it here: https://github.com/mcopik/hpx/blob/variadic_executor_params/hpx/parallel/executors/executor_parameters.hpp#L217

Maybe I have misunderstood variadic templates, but I'm afraid we can't use this traits type directly, right? We have to unpack and as a result we're going to obtain a unpacked sequence of booleans.

File: hpx/parallel/executors/auto_chunk_size.hpp (hunk 4 above)
Comment (mcopik) [inline hpx/parallel/executors/auto_chunk_size.hpp]: Yes, but I can't use traits directly, I think we have discussed already on Thursday afternoon? I merged variable_chunk_size and get_chunk_size into one executor parameter type called chunk_size, hence this function has to be defined for unwrapper. Otherwise I'd need again another tag for variable_chunk_size.

File: hpx/parallel/executors/executor_parameters.hpp (hunk 5 above)
Comment (mcopik) [inline hpx/parallel/executors/executor_parameters.hpp]: Sorry, I have forgot about that - fixed.

File: hpx/parallel/executors/executor_parameters.hpp (hunk 6 above)
Comment (mcopik) [inline hpx/parallel/executors/executor_parameters.hpp]: Fixed!

File: hpx/traits/is_executor_parameters.hpp (hunk 8 above)
Comment (mcopik) [inline hpx/traits/is_executor_parameters.hpp]: Maybe the name is confusing, I changed it to have a general scheme for all types. I'm not sure, should I change the name to emphasize how we use detailed implementation?

Yes, we can change it, but it's not necessary - is_executor_parameters inherits from detailed version using both template types.

File: hpx/traits/is_executor_parameters.hpp (hunk 9 above)
Comment (mcopik) [inline hpx/traits/is_executor_parameters.hpp]: If anyone is reading that and wondering: we have discussed it already with Hartmut outside GitHub. So far we haven't found better solution.

File: hpx/parallel/executors/executor_parameters.hpp (hunk 7 above)
Comment (mcopik) [inline hpx/parallel/executors/executor_parameters.hpp]: The test has been already implemented in HPX.

It seems I'm no setting a precedence here:

> grep -e "nullptr" -nrw .
> ./hpx/runtime/threads/policies/throttle_queue_scheduler.hpp:117:            nanosleep(&tim, nullptr);
> ./hpx/compute/vector.hpp:140:            if(static_cast<T*>(data_) != nullptr)
> ./hpx/compute/vector.hpp:156:            if(static_cast<T*>(data_) != nullptr)
> ./hpx/compute/detail/iterator.hpp:56:          : p_(nullptr)
> ./hpx/compute/detail/iterator.hpp:57:          , target_(nullptr)
> ./hpx/compute/cuda/target_ptr.hpp:43:          : p_(nullptr)
> ./hpx/compute/cuda/target_ptr.hpp:44:          , tgt_(nullptr)

Grep done a one recent checkout from hpx's master. I'm willing to add this one liner, should I change those instances too? Should I make another pull request - it's a small change, but completely unrelated to variadic executor parameters.
